src/stream/src/executor/top_n/group_top_n_eowc.rs

```diff
@@ -0,0 +1,419 @@
+                {
+                    let counts: Vec<i64> = counts_row
+                        .iter()
+                        .map(|v| v.map_or(0, ScalarRefImpl::into_int64))
+                        .collect();
+                    debug_assert_eq!(counts.len(), 1);
+                    counts[0]
+                } else {
+                    // ensure there is a row in the table for this group key
+                    self.group_key_table
+                        .insert((&group_key).chain(row::repeat_n(Some(ScalarImpl::from(0i64)), 1)));
+                    0
+                };
+                self.cache.put(compacted_key.clone(), count); // TODO(rc): can we avoid this clone?
+                self.cache.get_mut(&compacted_key).unwrap()
+            };
+
+            // snapshot the counts as prev counts when first time seeing this group key
+            prev_count_map
+                .entry(group_key.to_owned_row())
+                .or_insert_with(|| count.to_owned());
+
+            match op {
+                Op::Insert | Op::UpdateInsert => {
+                    *count += 1;
+                }
+                Op::Delete | Op::UpdateDelete => {
+                    *count -= 1;
+                    debug_assert!(*count >= 0);
+                }
+            }
+        }
+
+        // flush changes to state table
+        prev_count_map
+            .into_iter()
+            .for_each(|(group_key, prev_count)| {
+                let new_count = *self
+                    .cache
+                    .get(&CompactedRow::from(&group_key)) // TODO(rc): is it necessary to avoid recomputing here?
+                    .expect("group key in `prev_counts_map` must also exist in `self.cache`");
+                let new_count = OwnedRow::new(vec![Some(new_count.into())]);
+                let old_count = OwnedRow::new(vec![Some(prev_count.into())]);
+                self.group_key_table.update(
+                    group_key.clone().chain(old_count),
+                    group_key.chain(new_count),
+                );
+            });
+
+        // if we determine to flush to the table when processing every chunk instead of barrier
+        // coming, we can evict all including current epoch data.
+        self.cache.evict();
+
+        Ok(())
+    }
+
+    /// Flush the deduplication table.
+    fn flush(&mut self) {
+        // TODO(rc): now we flush the table in `dedup` method.
+        // WARN: if you want to change to batching the write to table. please remember to change
+        // `self.cache.evict()` too.
+        self.cache.evict();
+    }
+
+    /// Consume all group keys under `watermark`.
+    #[try_stream(ok = OwnedRow, error = StreamExecutorError)]
+    pub async fn consume<'a>(&'a mut self, watermark: ScalarImpl) {
+        #[for_await]
+        for row in self
+            .sort_buffer
+            .consume(watermark, &mut self.group_key_table)
+        {
+            yield row?;
+        }
+    }
```

Comment:
Maybe we can delete it in the LRU cache, or just like now wait for it to be evicted naturally.